File: app/src/main/java/com/android/ashwiask/tvmaze/favorite/FavoriteShowsRepository.java
package com.android.ashwiask.tvmaze.favorite;

import com.android.ashwiask.tvmaze.db.favouriteshow.FavoriteShow;
import com.android.ashwiask.tvmaze.db.favouriteshow.ShowDao;
import com.android.ashwiask.tvmaze.home.Show;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Single;

@Singleton
public class FavoriteShowsRepository {
    private final ShowDao showDao;

    @Inject
    public FavoriteShowsRepository(ShowDao showDao) {
        this.showDao = showDao;
    }

    public Single<List<FavoriteShow>> getAllFavoriteShows() {
        return showDao.getAllFavouriteShows();
    }

    public boolean isFavoriteShow(long id) {
        return showDao.isFavouriteShow(id) > 0;
    }

    public void insertShowIntoFavorites(Show show) {
        FavoriteShow favoriteShow = new FavoriteShow();
        favoriteShow.setId(show.id());
        favoriteShow.setImageUrl(show.image().get("original"));
        favoriteShow.setRating(show.rating().get("average"));
        favoriteShow.setName(show.name());
        favoriteShow.setPremiered(show.premiered());
        favoriteShow.setSummary(show.summary());
        favoriteShow.setFavorite(true);
        showDao.insert(favoriteShow);
    }

    public void removeShowFromFavorites(Show show) {
        FavoriteShow favoriteShow = new FavoriteShow();
        favoriteShow.setId(show.id());
        favoriteShow.setImageUrl(show.image().get("original"));
        favoriteShow.setRating(show.rating().get("average"));
        favoriteShow.setName(show.name());
        favoriteShow.setPremiered(show.premiered());
        favoriteShow.setSummary(show.summary());
        favoriteShow.setFavorite(false);
        showDao.remove(favoriteShow);
    }

    public void insertIntoFavorites(FavoriteShow favoriteShow) {
        showDao.insert(favoriteShow);
    }

    public void removeFromFavorites(FavoriteShow favoriteShow) {
        showDao.remove(favoriteShow);
    }
}

File: app/src/main/java/com/android/ashwiask/tvmaze/favorite/FavoriteShowsRepository.kt
package com.android.ashwiask.tvmaze.favorite

import com.android.ashwiask.tvmaze.db.favouriteshow.FavoriteShow
import com.android.ashwiask.tvmaze.db.favouriteshow.ShowDao
import com.android.ashwiask.tvmaze.network.home.Show
import io.reactivex.Single

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FavoriteShowsRepository @Inject
constructor(private val showDao: ShowDao) {

    val allFavoriteShows: Single<List<FavoriteShow>>
        get() = showDao.allFavouriteShows()

    fun insertShowIntoFavorites(show: Show) {
        val favoriteShow = FavoriteShow(
            id = show.id,
            name = show.name,
            premiered = show.premiered,
            imageUrl = show.image!!["original"],
            summary = show.summary,
            rating = show.rating!!["average"],
            runtime = show.runtime!!,
            isFavorite = true
        )
        showDao.insert(favoriteShow)
    }

    fun removeShowFromFavorites(show: Show) {
        val favoriteShow = FavoriteShow(
            id = show.id,
            name = show.name,
            premiered = show.premiered,
            imageUrl = show.image!!["original"],
            summary = show.summary,
            rating = show.rating!!["average"],
            runtime = show.runtime!!,
            isFavorite = false
        )
        showDao.remove(favoriteShow)
    }

    fun insertIntoFavorites(favoriteShow: FavoriteShow) {
        showDao.insert(favoriteShow)
    }

    fun removeFromFavorites(favoriteShow: FavoriteShow) {
        showDao.remove(favoriteShow)
    }

    val allFavoriteShowIds: Single<List<Long>>
        get() = showDao.getFavoriteShowIds()
}
